### ISSUE TITLE
upsample mask in segment/predict.py 

### ISSUE THREAD

**Issue by nikky4D** [ultralytics/yolov5#10297]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

how do i upsample the mask produced in segment/predict.py back to initial image shape? I modified the predict.py code as such, but the mask I am getting is not the correct shape of the input image

At line 153: , I add these lines
https://github.com/ultralytics/yolov5/blob/31c1f111868fc0dd7140ddce13e743f79bfaa9d4/segment/predict.py#L153
```


if len(det):
                masks = process_mask(proto[i], det[:, 6:], det[:, :4], im.shape[2:], upsample=True)  # HWC
                det[:, :4] = scale_boxes(im.shape[2:], det[:, :4], im0.shape).round()  # rescale boxes to im0 size
                
                ## Find the max, get max confidence, and save the mask of the max confidence
                max_loc = torch.argmax(det[:,4], axis=0)
                max_conf = det[max_loc,4].detach().cpu().numpy()
                conf_str = f"{max_conf:0.2f}"
                
                ## Save mask
                save_path_mask = str(save_dir / p.stem) + "_" + str(frame_num) + "conf_" + conf_str + "_mask.png"
                mask_i = masks[max_loc,...].detach().cpu().numpy()
                maskimg= PIL.Image.fromarray((mask_i*255).astype(np.uint8).squeeze(), 'L')
                maskimg.save(save_path_mask, "PNG")

```

I don't quite know how to upsample the mask shape back to original image shape im0.shape. Any help is appreciated

**Comment by glenn-jocher** [comment]
@nikky4D segment/predict.py does this automatically, you don't need to modify any code.

**Comment by nikky4D** [comment]
I do not see that, unless I'm mistaken. Printing out the shapes I get:
```
 for path, im, im0s, vid_cap, s in dataset:
        print(f"Preprocessed Image shape: {im.shape}")
        print(f"Orig Image shape: {im0s.shape}")
        print(f"Image path {path}")

...etc...

if len(det):
                masks = process_mask(proto[i], det[:, 6:], det[:, :4], im.shape[2:], upsample=True)  # HWC
                det[:, :4] = scale_boxes(im.shape[2:], det[:, :4], im0.shape).round()  # rescale boxes to im0 size
                
                ## Find the max, get max confidence, and save the mask of the max confidence
                max_loc = torch.argmax(det[:,4], axis=0)
                max_conf = det[max_loc,4].detach().cpu().numpy()
                conf_str = f"{max_conf:0.2f}"
                # print(det[max_loc,:5])
                # print(masks.shape)
                # print(p.stem)
                save_path_mask = str(save_dir / p.stem) + "_" + str(frame_num) + "conf_" + conf_str + "_mask.png"
                mask_i = masks[max_loc,...].detach().cpu().numpy()
                maskimg= PIL.Image.fromarray((mask_i*255).astype(np.uint8).squeeze(), 'L')
                maskimg.save(save_path_mask, "PNG")
                print(f"Mask shape: {masks.shape}")
                print(f"Det shape: {det.shape}")
                print(f"Orig Image shape: {im0s.shape}")
                wbboxes = det[:, :5].detach().cpu().numpy()
                box_paths = ["N/A"]*len(det)
                mask_paths = ["n/a"]*len(det)
                mask_paths[max_loc] = save_path_mask

```

Preprocessed Image shape: (3, 384, 640)
Orig Image shape: (1080, 1920, 3)
Mask shape: torch.Size([2, 384, 640])
Det shape: torch.Size([2, 38])
Orig Image shape: (1080, 1920, 3)

The mask shape is not the original image shape. How would I set it back to the original image shape.

**Comment by glenn-jocher** [comment]
@nikky4D segment/predict.py does this within the for loop.

**Comment by nikky4D** [comment]
Could you point me to where this is done? I am trying to pull out the binary mask of the detected object class (here human, class 0) and save that.

**Comment by glenn-jocher** [comment]
https://github.com/ultralytics/yolov5/blob/350e8eb69e01bb162ec0b22d1d13a1d1c2752853/segment/predict.py#L153-L171

**Comment by nikky4D** [comment]
The scaling up to original image shape does not occur at line 154. I put print statements after line 154, and the mask shape matched the shape of im, the preprocessed image,, but not of im0, the original input image. 

Is it the scale_segments section at line 159 also scale the mask? How would i then get out a mask from this?

**Comment by UNeedCryDear** [comment]
under this line https://github.com/ultralytics/yolov5/blob/350e8eb69e01bb162ec0b22d1d13a1d1c2752853/utils/plots.py#L161
add a return for functions named masks : ```return masks```<br>
and then,add code for annotator.masks
https://github.com/ultralytics/yolov5/blob/350e8eb69e01bb162ec0b22d1d13a1d1c2752853/segment/predict.py#L168
```
                maskssss=annotator.masks(masks,
                                colors=[colors(x, True) for x in det[:, 5]],
                                im_gpu=None if retina_masks else im[i])
                for index in range(len(maskssss)):
                    cv2.imshow(names[int(det[index,6])],maskssss[index].cpu().numpy())
                    cv2.waitKey()
```

**Comment by nikky4D** [comment]
Thank you. Is there an advantage to this way vs this modification I made: I wanted to avoid drawing the annotator masks and instead just upsample the mask using the scaled detection box
```
            if len(det):
                print(f"Preprocessed Image shape: {im.shape}")
                # masks = process_mask(proto[i], det[:, 6:], det[:, :4], im.shape[2:], upsample=True)  # HWC of im
                det[:, :4] = scale_boxes(im.shape[2:], det[:, :4], im0.shape).round()  # rescale boxes to im0 size
                masks = process_mask(proto[i], det[:, 6:], det[:, :4], im0.shape[:2], upsample=True)  # HWC of im0
```

**Comment by UNeedCryDear** [comment]
> Thank you. Is there an advantage to this way vs this modification I made: I wanted to avoid drawing the annotator masks and instead just upsample the mask using the scaled detection box

As long as you can correctly run your modification, it is no problem to do so.

In C++, I crop the feature map of mask-proto by position of the detection-box, and then calculate the mask results in the detection-box. This is much faster than directly calculating the whole image-segment-mask and then calculating the area in the detection-box, especially when the number of detections is large. However, Python's for-loop is inefficient, and the speed improvement is not obvious. But you can try it.</br>
picture for a Soul Painter:
![Soul Painter](https://user-images.githubusercontent.com/52729998/205261918-28a783ed-9af4-4900-adb8-fc3aac1169c7.png)</br>

**Comment by nikky4D** [comment]
Thanks for the drawing. I'll attempt it and see if it improves my timing

**Comment by glenn-jocher** [comment]
@nikky4D you're welcome! Feel free to give it a try, and let me know if you have any more questions along the way. Good luck with your implementation!
